docs/proposals/007-variadic-generics.md

```diff
@@ -0,0 +1,662 @@
+
+```
+// The callback function `p` is represented as a functor conforming to the `IProcessor` interface.
+void process<TProcessor:IFunc<void, int>>(TProcessor p, int data[N])
+{
+    for (int i = 0; i < N; i++)
+        p.process(data[i]);
+}
+
+void user(int myData[N])
+{
+    process((int x)=>{...}, myData);
+}
+```
+
+Related Work
+------------
+
+Variadic generics is an advance type system feature that is missinng in many modern languages including C# and Rust.
+Swift adds support for variadic generics recently in late 2022/2023, and this proposal largely follows Swift's design.
+C++ has variadic templates that achieves similar results within the template system.
+Rust supports variadics in a macro system layered above its core type system. While this can solve many user issues,
+we decided to not go through this path because macros and templates must be expanded before core type checking, which means that
+they can't integrate nicely in modules and compiled into IR independently of their use sites.
+
+
+Proposed Approach
+-----------------
+
+Slang can follow Swift's solution for variadic generics. A user can define a variadic generic with the syntax:
+
+```
+void myFunc<each T>(expand each T v) {...}
+```
+
+The code above defines a generic function that has a __generic type pack parameter__ `T` with the `each` keyword before `T`.
+The function's parameter list is defined as `expand each T v`, which should be interpreted as a parameter `v` whose type is
+`expand each T`. `expand each T` is a type that represents a pack of types. A parameter whose type is a pack of types can
+accept zero or more arguments during function call resolution.
+
+`myFunc` can be called with arbitrary number of arguments:
+```
+myFunc(); // OK, zero arguments
+myFunc(1, 2.0f, 3.0h); // OK, three arguments with different types.
+```
+
+A function can forward its variadic parameter to another function that accepts variadic parameter with the `expand` expression:
+```
+void caller<each T>(expand each T v)
+{
+    myFunc(expand each v);
+}
+```
+
+Generic type pack parameters can be nested, and there can be more than one variadic generic parameters in a single generic decl:
+```
+struct Parent<each T>
+{
+    void f<each U>(...) {...} // OK, nested generics with type pack parameters
+}
+void g<each T, each U>(...) { ... } // OK, more than one type pack parameter in a single generic.
+```
+
+However, when more than one generic type pack parameters is referenced in a single `expand` expression, there is an implicit
+requirement that these type packs will have the same number of elements. For example:
+```
+ // implicitly requiring T and U to have same number of elements.
+void g<each T, each U>(expand Pair<each T, each U> pairs) {...}
+
+void user()
+{
+    // We will match (int, float) to `T`, and (uint16_t, half) to `U`:
+    g<int, float, uint16_t, half>(
```

Comment:
what happens for 
`g<int, float, uint16_t, half, int>`?

Will it report error?

Reply:
Yes, that will be an error.